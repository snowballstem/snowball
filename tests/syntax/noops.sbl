// Test no-op commands which should get canonicalised.
externals (stem)
integers (x y)
booleans (b)
strings (s)
define stem as (
  $x = 0
  [substring] among ( // All actions equivalent to `true`.
    '0' ()
    'a' (())
    'b' (true)
    'c' ('')
    'd' (loop 0 next)
    'e' (loop -1 next)
    'f' (hop 0)
    'g' ($x += 0)
    'h' ($x -= 0)
    'i' ($x *= 1)
    'j' ($x /= 1)

    // Values set are unused.
    'k' ($y = 1)
    'l' ($y += 1)
    'm' ($y -= 1)
    'n' ($y *= 2)
    'o' ($y /= 2)
    'p' (setmark y)
    'q' (set b)
    'r' (unset b)
    's' (-> s)

    // Integer tests.
    'A' ($(1 == 1))
    'B' ($(2 > 1))
    'C' ($(-1 < 0))
    'D' ($(-1 <= -1))
    'E' ($(0 >= 0))
    'F' ($(1 != 0))
  )
  [substring] among ( // All actions equivalent to `false`.
    '0' (false)
    'a' (fail ())
    'b' (fail true)
    'c' (fail '')
    'd' (fail loop 0 next)
    'e' (fail loop -1 next)
    'f' (fail hop 0)
    'g' (fail $x += 0)
    'h' (fail $x -= 0)
    'i' (fail $x *= 1)
    'j' (fail $x /= 1)

    // Integer tests which are always false.
    'A' ($(1 != 1))
    'B' ($(1 > 2))
    'C' ($(-1 > 0))
    'D' ($(0 <= -1))
    'E' ($(0 >= 1))
    'F' ($(1 == 0))
  )
)

