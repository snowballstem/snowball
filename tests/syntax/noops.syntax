define stem
  (
    true
    [
    substring
    ]
    among
      literal '0'
      (
      literal 'a'
      (
      literal 'b'
      (
      literal 'c'
      (
      literal 'd'
      (
      literal 'e'
      (
      literal 'f'
      (
      literal 'g'
      (
      literal 'h'
      (
      literal 'i'
      (
      literal 'j'
      (
      literal 'k'
      (
      literal 'l'
      (
      literal 'm'
      (
      literal 'n'
      (
      literal 'o'
      (
      literal 'p'
      (
      literal 'q'
      (
      literal 'r'
      (
      literal 's'
      (
      literal 'A'
      (
      literal 'B'
      (
      literal 'C'
      (
      literal 'D'
      (
      literal 'E'
      (
      literal 'F'
      (
    [
    substring
    ]
    among
      literal '0'
      (
        false
      literal 'a'
      (
        false
      literal 'b'
      (
        false
      literal 'c'
      (
        false
      literal 'd'
      (
        false
      literal 'e'
      (
        false
      literal 'f'
      (
        false
      literal 'g'
      (
        false
      literal 'h'
      (
        false
      literal 'i'
      (
        false
      literal 'j'
      (
        false
      literal 'A'
      (
        false
      literal 'B'
      (
        false
      literal 'C'
      (
        false
      literal 'D'
      (
        false
      literal 'E'
      (
        false
      literal 'F'
      (
        false
