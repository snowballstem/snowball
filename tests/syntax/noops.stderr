syntax/noops.sbl:12: warning: empty literal string is a no-op
syntax/noops.sbl:13: warning: loop 0 C is a no-op
syntax/noops.sbl:14: warning: loop -1 C is a no-op
syntax/noops.sbl:15: warning: hop 0 is a no-op
syntax/noops.sbl:44: warning: empty literal string is a no-op
syntax/noops.sbl:45: warning: loop 0 C is a no-op
syntax/noops.sbl:46: warning: loop -1 C is a no-op
syntax/noops.sbl:47: warning: hop 0 is a no-op
syntax/noops.sbl:5: warning: string 's' is set but never used
syntax/noops.sbl:4: warning: boolean 'b' is set but never used
syntax/noops.sbl:3: warning: integer 'y' is set but never used
syntax/noops.sbl:3: warning: integer 'x' is set but never used
